File: funcscript/funcs/list/DistinctListFunction.cs
using funcscript.core;
using funcscript.model;

namespace funcscript.funcs.list
{
    public class DistinctListFunction : IFsFunction
    {
        public int MaxParsCount => 1;

        public CallType CallType => CallType.Prefix;

        public string Symbol => "Distinct";

        public int Precidence => 0;

        class DistinctListFuncPar : IParameterList
        {
            public object X;

            public int Count => 1;

            public object GetParameter(IFsDataProvider provider, int index)
            {
                return index switch
                {
                    0 => X,
                    _ => null,
                };
            }
        }

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            if (pars.Count != this.MaxParsCount)
                throw new error.EvaluationTimeException($"{this.Symbol} function: Invalid parameter count. Expected {this.MaxParsCount}, but got {pars.Count}");

            var par0 = pars.GetParameter(parent, 0);

            if (par0 == null)
                return null;

            if (!(par0 is FsList))
                throw new error.TypeMismatchError($"{this.Symbol} function: The parameter should be {this.ParName(0)}");

            var lst = (FsList)par0;

            var distinctValues = new HashSet<object>();
            var res = new List<object>();

            for (int i = 0; i < lst.Length; i++)
            {
                if (distinctValues.Add(lst[i]))
                {
                    res.Add(lst[i]);
                }
            }

            return new ArrayFsList(res);
        }

        public string ParName(int index)
        {
            if (index == 0)
                return "List";
            return "";
        }
    }
}

File: js_port/fs_js/src/funcs/list/DistinctListFunction.js
const { BaseFunction, CallType } = require('../../core/functionBase');
const { ArrayFsList } = require('../../model/FsList');
const helpers = require('../helpers');
const { FSDataType } = require('../../core/fstypes');

class DistinctListFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'Distinct';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 1;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, this.maxParameters);
    if (error) {
      return error;
    }

    const list = helpers.ensureList(parameters.getParameter(provider, 0));
    if (!list) {
      return helpers.makeError(
        helpers.FsError.ERROR_TYPE_MISMATCH,
        `${this.symbol} function: The parameter should be ${this.parName(0)}`
      );
    }

    const seen = [];
    const unique = [];
    for (let i = 0; i < list.length; i += 1) {
      const item = list.get(i);
      if (!seen.some((existing) => helpers.typeOf(existing) === helpers.typeOf(item) && helpers.valueOf(existing) === helpers.valueOf(item))) {
        seen.push(item);
        unique.push(item);
      }
    }

    return helpers.makeValue(FSDataType.List, new ArrayFsList(unique));
  }

  parName(index) {
    return index === 0 ? 'List' : '';
  }
}

module.exports = {
  DistinctListFunction
};
